File: backend/app/schemas/interview.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel

from app.core.enums import Difficulty, InterviewType, SessionStatus


class CreateSessionRequest(BaseModel):
    interview_type: InterviewType
    difficulty: Difficulty
    resume_id: uuid.UUID | None = None


class QuestionResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: uuid.UUID
    sequence_order: int
    question_text: str
    question_type: str
    created_at: datetime


class SessionResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: uuid.UUID
    interview_type: InterviewType
    difficulty: Difficulty
    status: SessionStatus
    started_at: datetime
    completed_at: datetime | None
    overall_score: float | None


class SessionDetailResponse(SessionResponse):
    questions: list[QuestionResponse] = []


class SubmitAnswerRequest(BaseModel):
    answer_text: str
    time_taken_seconds: int | None = None


class FeedbackResponse(BaseModel):
    model_config = {"from_attributes": True}
    overall_score: float
    feedback_text: str
    strengths: list[str]
    weaknesses: list[str]
    suggestions: list[str]
    technical_score: float | None
    communication_score: float | None
    correctness_score: float | None


class AnswerFeedbackResponse(BaseModel):
    answer_id: uuid.UUID
    feedback: FeedbackResponse


class CompleteSessionResponse(BaseModel):
    session_id: uuid.UUID
    overall_score: float
    questions_answered: int


class QuestionFeedbackItem(BaseModel):
    question: QuestionResponse
    answer_text: str | None
    feedback: FeedbackResponse | None


class SessionFeedbackResponse(BaseModel):
    session_id: uuid.UUID
    overall_score: float | None
    per_question: list[QuestionFeedbackItem]
    strengths: list[str]
    weaknesses: list[str]

File: frontend/src/types/interview.ts
// ─── Shared ─────────────────────────────────────────────────────────────────

export interface PageResponse<T> {
  items: T[]
  total: number
  limit: number
  offset: number
}

// ─── Enums (mirror backend app/core/enums.py) ──────────────────────────────

export type InterviewType = 'swe' | 'ml' | 'behavioral' | 'cybersecurity'
export type Difficulty = 'junior' | 'mid' | 'senior'
export type SessionStatus = 'active' | 'completed' | 'abandoned'

// ─── Request shapes ─────────────────────────────────────────────────────────

export interface CreateSessionRequest {
  interview_type: InterviewType
  difficulty: Difficulty
  resume_id?: string | null
}

export interface SubmitAnswerRequest {
  answer_text: string
  time_taken_seconds?: number | null
}

// ─── Response shapes ────────────────────────────────────────────────────────

/** Returned by GET /interviews/sessions (list item) and POST /interviews/sessions */
export interface Session {
  id: string
  interview_type: InterviewType
  difficulty: Difficulty
  status: SessionStatus
  started_at: string // ISO 8601
  completed_at: string | null
  overall_score: number | null
}

export interface Question {
  id: string
  sequence_order: number
  question_text: string
  question_type: string
  created_at: string // ISO 8601
}

/** Returned by GET /interviews/sessions/{id} */
export interface SessionDetail extends Session {
  questions: Question[]
}

export interface Feedback {
  overall_score: number
  feedback_text: string
  strengths: string[]
  weaknesses: string[]
  suggestions: string[]
  technical_score: number | null
  communication_score: number | null
  correctness_score: number | null
}

/** Returned by POST /interviews/sessions/{id}/questions/{qid}/answers */
export interface AnswerFeedback {
  answer_id: string
  feedback: Feedback
}

/** Returned by POST /interviews/sessions/{id}/complete */
export interface CompleteSessionResult {
  session_id: string
  overall_score: number
  questions_answered: number
}

export interface QuestionFeedbackItem {
  question: Question
  answer_text: string | null
  feedback: Feedback | null
}

/** Returned by GET /interviews/sessions/{id}/feedback */
export interface SessionFeedback {
  session_id: string
  overall_score: number | null
  per_question: QuestionFeedbackItem[]
  strengths: string[]
  weaknesses: string[]
}
